### ISSUE TITLE
[clang-tidy] `misc-const-correctness` false positive when returning a pointer to non-const type

### ISSUE THREAD

**Issue by palapapa** [llvm/llvm-project#199334]
```c++
auto foo() -> int* {
    static int *ptr = nullptr;
    return ptr;
}
```

This reports: `Pointee of variable 'ptr' of type 'int *' can be declared 'const'`, but changing it to `static const int *ptr` is an error because the function's return type requires a pointer to a non-const `int`.

LLVM version: 21.1.8

Clang-Tidy config:

```yaml
Checks:
  - "*"
  - -altera-*
  - -android-cloexec-*
  - -llvmlibc-*
  - -fuchsia-*
  - -readability-identifier-length
CheckOptions:
  modernize-use-default-member-init.UseAssignment: true
  readability-implicit-bool-conversion.AllowPointerConditions: true
  modernize-use-trailing-return-type.TransformLambdas: none
```

**Comment by berkaysahiin** [comment]
Doesn't seem to appear on trunk: https://godbolt.org/z/qxY7osx33

**Comment by palapapa** [comment]
> Doesn't seem to appear on trunk: https://godbolt.org/z/qxY7osx33

It also doesn't show when I switched to Clang 21.1.0, but it definitely shows on my machine.

**Comment by vbvictor** [comment]
If this doesn't repro on trunk, we should close the issue.
This could be already fixed by some patch in 22 or upcoming 23 release

**Comment by vbvictor** [comment]
Probably fixed by:
"Fixed false positives of function pointer, on return of non-const pointer, on pointer-to-member operator."
From https://releases.llvm.org/22.1.0/tools/clang/tools/extra/docs/ReleaseNotes.html

**Comment by alexolog** [comment]
I am also seeing it on homebrew clang 21.0.8
Could it be a regression?
We can't switch to 22 at this point

**Comment by vbvictor** [comment]
> I am also seeing it on homebrew clang 21.0.8
> Could it be a regression?
> We can't switch to 22 at this point

We don't support 21.x release at this point, only 22.x (and upcoming 23)
